// html-validate/dist/cjs/core-browser.js
'use strict';

var core = require('./core.js');

function importFunction(id) {
  return import(id);
}

async function internalImport(id) {
  const { default: defaultImport } = await importFunction(id);
  if (!defaultImport) {
    throw new core.UserError(`"${id}" does not have a default export`);
  }
  return defaultImport;
}
function esmResolver() {
  return {
    name: "esm-resolver",
    resolveElements(id) {
      return internalImport(id);
    },
    resolveConfig(id) {
      return internalImport(id);
    },
    resolvePlugin(id) {
      return internalImport(id);
    },
    async resolveTransformer(id) {
      return internalImport(id);
    }
  };
}

const defaults = {
  silent: false,
  version: core.version,
  logger(text) {
    console.error(text);
  }
};
function compatibilityCheck(name, declared, options) {
  return core.compatibilityCheckImpl(name, declared, {
    ...defaults,
    ...options
  });
}

exports.compatibilityCheck = compatibilityCheck;
exports.esmResolver = esmResolver;
